### PR TITLE
add notes and questions for n45

### PR DESCRIPTION
**Pull request by reardencode**

### CODE REVIEW

**Review by glozow — COMMENTED** [review]
Thanks @reardencode! This looks like a great start. You might want to add notes about the implementation, instructions for how people might test/use the code, and more description/links for somebody who isn't super familiar with the code.

File: _posts/2024-03-13-bitcoin-inquisition-#45.md [hunk 1]
```diff
@@ -11,17 +11,45 @@ status: upcoming
...
+This Pull Request adds a total of 3 opcodes:
+* `OP_INTERNALKEY` in Tapscript only, copies the Taproot Internal Key to the
+  stack.
+* `OP_CHECKSIGFROMSTACK` in Tapscript only, validates a BIP340 Schnorr
+  signature using an item from the stack as the message.
+* `OP_CHECKSIGFROMSTACKVERIFY` in all script types, validates a BIP340 Schnorr
+  signature using an item from the stack as the message, and leaves the stack
+  unchanged.
 
+The semantics of the signature checking opcodes are similar to Tapscript
```
Comment (LarryRuane) [inline _posts/2024-03-13-bitcoin-inquisition-#45.md:33]:
```suggestion
The semantics of the signature-checking opcodes are similar to Tapscript
```
silly nit, I know

File: _posts/2024-03-13-bitcoin-inquisition-#45.md [hunk 2]
```diff
@@ -11,17 +11,45 @@ status: upcoming
...
 ## Questions
 
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+
+## Open considerations
+
+1. Given callers of `OP_CHECKSIGFROMSTACK` can reasonably be expected to want
+   the ability to check a signature against multiple stack items, and that the
+   naieve implementation (using `OP_CAT <pubkey> OP_CHECKSIGFROMSTACK`) is
```
Comment (LarryRuane) [inline _posts/2024-03-13-bitcoin-inquisition-#45.md:45]:
```suggestion
   naive implementation (using `OP_CAT <pubkey> OP_CHECKSIGFROMSTACK`) is
```

**Review by LarryRuane — COMMENTED** [review]
I can review again after you add more text.

**Comment by reardencode** [comment]
Thanks both - added some additional text. Not sure what else might be helpful for review clubbers :)

File: _posts/2024-03-13-bitcoin-inquisition-#45.md [hunk 3]
```diff
@@ -11,17 +11,89 @@ status: upcoming
...
+I posted some scripts using `bitcoinjs-lib` and `noble-curves` that I used to
+generate the test vectors above:
+https://gist.github.com/reardencode/9dbc60a6d6e1591905d25bf4d123dfdd
 
 ## Questions
 
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+
+## Open considerations
```
Comment (glozow) [inline _posts/2024-03-13-bitcoin-inquisition-#45.md:85]: I'd recommend adding some questions to help people check their understanding. Softballs like "which nop is being repurposed" / "what does this opcode do" or conceptual questions "what's the motivation for this change" / "how is this useful" are good places to start. Otherwise the meeting will be very short...

File: _posts/2024-03-13-bitcoin-inquisition-#45.md [hunk 4]
```diff
@@ -11,17 +11,89 @@ status: upcoming
...
+### Testing
+
+For both INTERNALKEY and CSFS I wrote tests in the Transaction Test framework:
+
+https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/test/data/tx_valid.json#L671
+https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/test/data/tx_invalid.json#L491
+
+You can run a regtest node from this branch to use them yourself.
+
+I posted some scripts using `bitcoinjs-lib` and `noble-curves` that I used to
```
Comment (glozow) [inline _posts/2024-03-13-bitcoin-inquisition-#45.md:77]: Please use third person

File: _posts/2024-03-13-bitcoin-inquisition-#45.md [hunk 5]
```diff
@@ -11,17 +11,89 @@ status: upcoming
...
 ---
 
-_Notes and questions to follow soon!_
-
-<!-- TODO: Before meeting, add notes and questions
 ## Notes
 
+Attendees should familiarize themselves with
+[LNHANCE](https://delvingbitcoin.org/t/lnhance-bips-and-implementation/376),
+and [LN-Symmetry (Eltoo)](https://bitcoinops.org/en/topics/eltoo/).
```
Comment (glozow) [inline _posts/2024-03-13-bitcoin-inquisition-#45.md:18]: Maybe describe LNHANCE with a few sentences?

**Comment by reardencode** [comment]
TYVM @glozow updated per your review.

File: _posts/2024-03-13-bitcoin-inquisition-#45.md [hunk 6]
```diff
@@ -11,17 +11,106 @@ status: upcoming
...
+Outside of these changes in `interpreter.cpp`, the only non-flag change in
+production code is to allow variable length input to BIP340
+signature-checking.
+
+### Testing
+
+For both INTERNALKEY and CSFS I wrote tests in the Transaction Test framework:
+
+https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/test/data/tx_valid.json#L671
+https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/test/data/tx_invalid.json#L491
```
Comment (glozow) [inline _posts/2024-03-13-bitcoin-inquisition-#45.md:84]: raw links are not very reader-friendly
```suggestion
Test data for INTERNALKEY and CSFS are included in [tx_valid.json](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/test/data/tx_valid.json#L671) and [tx_invalid.json](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/test/data/tx_invalid.json#L491).
```

File: _posts/2024-03-13-bitcoin-inquisition-#45.md [hunk 7]
```diff
@@ -11,17 +11,106 @@ status: upcoming
...
+of `OP_INTERNALKEY` is remarkably simple. The taproot internal key is already
+extracted for use with verifying tapscript `OP_CHECKSIG(VERIFY|ADD)` with the
+key `0x01`, so it simply needs to be added to the stack.
+
+#### `OP_CHECKSIGFROMSTACK(VERIFY)`
 
+The implementation of CSFS(V) is split between the [main opcode
+processing](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/script/interpreter.cpp#L1318), and a function analogous to `EvalChecksigPreTapscript` and `EvalChecksigTapscript` named [`EvalChecksigFromStack`](https://github.com/bitcoin-core-review-club/bitcoin/blob/prbitcoin-inquisition-45/src/script/interpreter.cpp#L346).
+
+My original approach was to add a `SignatureChecker` that could carry the
```
Comment (glozow) [inline _posts/2024-03-13-bitcoin-inquisition-#45.md:63]: I still see first person here? You can say "a previous approach."

File: _posts/2024-03-13-bitcoin-inquisition-#45.md [hunk 8]
```diff
@@ -11,17 +11,106 @@ status: upcoming
...
 
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+1. What does `OP_CHECKSIGFROMSTACKVERIFY` leave on the stack?
+   `OP_CHECKSIGFROMSTACK`?
+1. In what case can `OP_CHECKSIGFROMSTACK` immediately fail?
+1. What length is the `data` argument to `OP_CHECKSIGFROMSTACK(VERIFY)`?
+1. How does `OP_INTERNALKEY OP_CHECKSIG` in Tapscript compare to key spend
+   in weight?
+
```
Comment (glozow) [inline _posts/2024-03-13-bitcoin-inquisition-#45.md:101]: nit: prefer having newlines in between questions

**Review by glozow — COMMENTED** [review]
Minor suggestions. I can also add a commit to push these up if you prefer.

File: _posts/2024-03-13-bitcoin-inquisition-#45.md [hunk 9]
```diff
@@ -11,17 +11,109 @@ status: upcoming
...
+3. In what case can `OP_CHECKSIGFROMSTACK` immediately fail?
+
+4. What length is the `data` argument to `OP_CHECKSIGFROMSTACK(VERIFY)`?
+
+5. How does `OP_INTERNALKEY OP_CHECKSIG` in Tapscript compare to key spend
+   in weight?
+
+## Open considerations
+
+1. Given callers of `OP_CHECKSIGFROMSTACK` can reasonably be expected to want
```
Comment (glozow) [inline _posts/2024-03-13-bitcoin-inquisition-#45.md:107]: This question was confusing for me, as I am missing explanations for:
- why callers would be "reasonably expected to want" this ability
- why `OP_CAT <pubkey> OP_CHECKSIGFROMSTACK` is "the naive implementation"
- why it is insecure (where is this explained?)

Comment (glozow) [reply _posts/2024-03-13-bitcoin-inquisition-#45.md:107]: My recommendation is to split this into multiple questions and have notes walking us through each assertion you're making

File: _posts/2024-03-13-bitcoin-inquisition-#45.md [hunk 10]
```diff
@@ -11,17 +11,109 @@ status: upcoming
...
-<!-- TODO: Before meeting, add notes and questions
 ## Notes
 
+Attendees should familiarize themselves with
+[LNHANCE](https://delvingbitcoin.org/t/lnhance-bips-and-implementation/376),
+and [LN-Symmetry (Eltoo)](https://bitcoinops.org/en/topics/eltoo/).
+
+TL;DR: LNHANCE is a soft fork proposal for bitcoin which combines 4 total
+opcodes: `OP_CHECKTEMPLATEVERIFY`, `OP_CHECKSIGFROMSTACK`,
+`OP_CHECKSIGFROMSTACKVERIFY`, and `OP_INTERNALKEY`. It consumes 2 NOPs in all
```
Comment (glozow) [inline _posts/2024-03-13-bitcoin-inquisition-#45.md:22]: clarification for myself: by "consumes," I assume you mean repurposes?

File: _posts/2024-03-13-bitcoin-inquisition-#45.md [hunk 11]
```diff
@@ -11,17 +11,109 @@ status: upcoming
...
+
+2. What does `OP_CHECKSIGFROMSTACKVERIFY` leave on the stack?
+   `OP_CHECKSIGFROMSTACK`?
+
+3. In what case can `OP_CHECKSIGFROMSTACK` immediately fail?
+
+4. What length is the `data` argument to `OP_CHECKSIGFROMSTACK(VERIFY)`?
+
+5. How does `OP_INTERNALKEY OP_CHECKSIG` in Tapscript compare to key spend
+   in weight?
```
Comment (glozow) [inline _posts/2024-03-13-bitcoin-inquisition-#45.md:103]: Some more question ideas
- Did you run the tests? Is there any missing test coverage? Can you think of ways to extend the tests?
- How might one construct a vault (or one of the examples you mention above)?
- What concrete improvements to LN can be achieved this way?

File: _posts/2024-03-13-bitcoin-inquisition-#45.md [hunk 12]
```diff
@@ -11,17 +11,109 @@ status: upcoming
...
+that approach had two problems: getting BIP340 signing into legacy/segwitv0
+scripts became intrusive, and the semantics of legacy/segwitv0 signature
+checking are not particularly desirable for a new opcode. As a result the
+current implementation uses a separate function specific to CSFS and gives
+modern Tapscript-style semantics. Only after changing code approaches did it
+become clear that ECDSA checking might not be desirable in these new sigops.
+The commit history still reflects the removal of ECDSA in a separate commit.
+
+Outside of these changes in `interpreter.cpp`, the only non-flag change in
+production code is to allow variable length input to BIP340
```
Comment (glozow) [inline _posts/2024-03-13-bitcoin-inquisition-#45.md:75]: Confused as to what you mean by "production code." Also I don't think the term "flag change" (and "non-flag change") is defined?

**Review by glozow — COMMENTED** [review]
I've added some notes on where you may need to elaborate on things during the meeting, as not everybody is following all the covenant discussion. Hopefully people will have attended the OP_CAT and/or read up on BIP 119 and vaults, but not everybody will. Just to give you a heads up.

I am fine with these notes, so will merge unless you want to change anything - let me know

**Comment by reardencode** [comment]
Thanks Gloria! Updated based on your suggestions and ready for merge IMO.

**Comment by glozow** [comment]
Thanks for taking suggestions @reardencode! I've added a commit with very minor changes (largely just clarifying abbreviations), and rfm.

**Review by glozow — APPROVED** [review]
